src/main/java/htsjdk/variant/vcf/VCFPathReader.java

```diff
@@ -0,0 +1,183 @@
+package htsjdk.variant.vcf;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.Interval;
+import htsjdk.samtools.util.IntervalList;
+import htsjdk.tribble.AbstractFeatureReader;
+import htsjdk.tribble.FeatureCodec;
+import htsjdk.tribble.FeatureReader;
+import htsjdk.tribble.TribbleException;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.variantcontext.VariantContext;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * Simplified interface for reading from VCF/BCF files.
+ */
+public class VCFPathReader implements Closeable, Iterable<VariantContext> {
+
+    private final FeatureReader<VariantContext> reader;
+
+    /**
+     * Returns true if the given path appears to be a BCF file.
+     */
+    public static boolean isBCF(final Path path) {
+        return path.toUri().getRawPath().endsWith(".bcf");
+    }
+
+    /**
+     * Returns the SAMSequenceDictionary from the provided VCF file.
+     */
+    public static SAMSequenceDictionary getSequenceDictionary(final Path path) {
+        final SAMSequenceDictionary dict = new VCFPathReader(path, false).getFileHeader().getSequenceDictionary();
+        CloserUtil.close(path);
+        return dict;
+    }
+
+    /**
+     * Constructs a VCFFileReader that requires the index to be present.
+     */
+    public VCFPathReader(final Path path) {
+        this(path, true);
+    }
+
+    /**
+     * Constructs a VCFFileReader with a specified index.
+     */
+    public VCFPathReader(final Path path, final Path indexPath) {
+        this(path, indexPath, true);
+    }
+
+    /**
+     * Allows construction of a VCFFileReader that will or will not assert the presence of an index as desired.
+     */
+    public VCFPathReader(final Path path, final boolean requireIndex) {
+        // Note how we deal with type safety here, just casting to (FeatureCodec)
+        // in the call to getFeatureReader is not enough for Java 8.
```

Comment:
This is because the compiler can't infer the common type that both branches of the ternary op produces and the function accepts. If you cast using the parameterized type you can inline this expression, e.g. `(FeatureCodec<VariantContext, ?>) (isBCF(path) ? new BCF2Codec() : new VCFCodec())` works.

Reply:
It looks like `isBCF()` is only called in two places and in both places it's used to determine which codec to return. If so you can avoid the code duplication and and replace with `isBCF()` with `getCodecForPath(Path)`. Or keep `isBCF()` and write:
```
    private static FeatureCodec<VariantContext, ?> getCodecForPath(Path path) {
        return isBCF(path) ? new BCF2Codec() : new VCFCodec();
    }
```
This also eliminates the need for a cast or introducing a local variable.